### PR TITLE
Added classes for synthetic test data creation

### PR DESCRIPTION
**Pull request by lorenz-gorini**
Added a new framework for synthetic data creation that can be used for testing the pytrousse FeatureOperation application.

The idea is to have a tool for simulating raw data starting from clean ones. Since some ReverseFeatureOperation cannot be perfectly reversed by applying pytrousse FeatureOperation, the simulated raw data must be created along with the data that the user will get after that the related FeatureOperation is applied.

### CODE REVIEW

File: tests/test_dataset.py [hunk 1]
```diff
@@ -0,0 +1,713 @@
...
+    ):
+        """
+        Insert NaN values into "self._correct_column" and "self._column_with_errors".
+
+        The function inserts a ``error_count`` number of
+        equally-spaced NaN values.
+
+        Parameters
+        ----------
+        column_names: Union[List[str], List[int]]
```
Comment (alessiamarcolini) [inline tests/test_dataset.py:494]:
```suggestion
        column_names : Union[List[str], List[int]]
```

File: tests/test_dataset.py [hunk 2]
```diff
@@ -0,0 +1,713 @@
...
+            List of the names/column IDs of the columns on which the
+            ReverseFeatureOperation is applied. A mix of names and column IDs
+            is not accepted.
+        error_count: int
```
Comment (alessiamarcolini) [inline tests/test_dataset.py:498]:
```suggestion
        error_count : int
```

File: tests/test_dataset.py [hunk 3]
```diff
@@ -0,0 +1,713 @@
...
+        """
+        super().__init__(column_names=column_names)
+        self._error_count = error_count
+
+    def __call__(self, dataset: TestDataSet) -> TestDataSet:
+        """
+        Insert NaN values into ``dataset`` keeping track of the modified samples.
+
+        The function inserts an ``error_count`` number of equally-spaced NaN values
+        into the ``self._columns`` attribute.
```
Comment (alessiamarcolini) [inline tests/test_dataset.py:509]: these implementation details shouln't be in the docstring

File: tests/test_dataset.py [hunk 4]
```diff
@@ -0,0 +1,713 @@
...
+
+        Parameters
+        ----------
+        dataset: TestDataSet
+            TestDataSet instance containing the values of a dataset where some
+            invalid substrings will be inserted in the columns ``self._columns``.
```
Comment (alessiamarcolini) [inline tests/test_dataset.py:515]: same here

File: tests/test_dataset.py [hunk 5]
```diff
@@ -0,0 +1,713 @@
...
+    def __call__(self, dataset: TestDataSet) -> TestDataSet:
+        """
+        Insert NaN values into ``dataset`` keeping track of the modified samples.
+
+        The function inserts an ``error_count`` number of equally-spaced NaN values
+        into the ``self._columns`` attribute.
+
+        Parameters
+        ----------
+        dataset: TestDataSet
```
Comment (alessiamarcolini) [inline tests/test_dataset.py:513]:
```suggestion
        dataset : TestDataSet
```

File: tests/test_dataset.py [hunk 6]
```diff
@@ -0,0 +1,713 @@
...
+        Insert invalid strings into the column and store the related correct values.
+
+        The function inserts an ``error_count`` number of
+        equally-spaced invalid strings into the column. These strings are create
+        by using ``replacement_map`` dictionary that connects the original
+        substring with the invalid one that is replaced.
+
+        Parameters
+        ----------
+        column_names: Union[List[str], List[int]]
```
Comment (alessiamarcolini) [inline tests/test_dataset.py:562]:
```suggestion
        column_names : Union[List[str], List[int]]
```

File: tests/test_dataset.py [hunk 7]
```diff
@@ -0,0 +1,713 @@
...
+            List of the names/column IDs of the columns on which the
+            ReverseFeatureOperation is applied. A mix of names and column IDs
+            is not accepted.
+        error_count: int
```
Comment (alessiamarcolini) [inline tests/test_dataset.py:566]:
```suggestion
        error_count : int
```

File: tests/test_dataset.py [hunk 8]
```diff
@@ -0,0 +1,713 @@
...
+            Number of invalid strings that are inserted in column.
+        replacement_map: Dict
```
Comment (alessiamarcolini) [inline tests/test_dataset.py:568]:
```suggestion
        replacement_map : Dict
```

File: tests/test_dataset.py [hunk 9]
```diff
@@ -0,0 +1,713 @@
...
+        Insert invalid strings into the column and store the related correct values.
+
+        The function inserts an ``error_count`` number of
+        equally-spaced invalid strings into the column. These strings are create
+        by using ``replacement_map`` dictionary that connects the original
+        substring with the invalid one that is replaced.
+
+        Parameters
+        ----------
+        dataset: TestDataSet
```
Comment (alessiamarcolini) [inline tests/test_dataset.py:589]:
```suggestion
        dataset : TestDataSet
```

File: tests/test_dataset.py [hunk 10]
```diff
@@ -0,0 +1,713 @@
...
+
+        The function inserts an ``error_count`` number of
+        equally-spaced invalid strings into the column. These strings are created
+        by using ``replacement_map`` dictionary that connects the original
+        substring with the invalid one that is replaced, making some column values
+        not coherent with the others. E.g. 08/03/2020 -> 08-03-2020
+
+        Parameters
+        ----------
+        column_names: Union[List[str], List[int]]
```
Comment (alessiamarcolini) [inline tests/test_dataset.py:645]:
```suggestion
        column_names : Union[List[str], List[int]]
```

File: tests/test_dataset.py [hunk 11]
```diff
@@ -0,0 +1,713 @@
...
+            List of the names/column IDs of the columns on which the
+            ReverseFeatureOperation is applied. A mix of names and column IDs
+            is not accepted.
+        error_count: int
```
Comment (alessiamarcolini) [inline tests/test_dataset.py:649]:
```suggestion
        error_count : int
```

File: tests/test_dataset.py [hunk 12]
```diff
@@ -0,0 +1,713 @@
...
+            Number of errors that are inserted in column.
+        replacement_map: Dict
```
Comment (alessiamarcolini) [inline tests/test_dataset.py:651]:
```suggestion
        replacement_map : Dict[str, str]
```

File: tests/test_dataset.py [hunk 13]
```diff
@@ -0,0 +1,713 @@
...
+        self._validate_replacement_map(replacement_map)
+        self._replacement_map = replacement_map
+
+    def _validate_replacement_map(self, replacement_map: Dict[str, str]):
+        """
+        Validate ``replacement_map`` argument used to initialize the instances
+
+        Parameters
+        ----------
+        replacement_map: Dict
```
Comment (alessiamarcolini) [inline tests/test_dataset.py:668]:
```suggestion
        replacement_map : Dict[str, str]
```

File: tests/test_dataset.py [hunk 14]
```diff
@@ -0,0 +1,713 @@
...
+        """
+        Insert invalid substrings into the ``column`` and store the related correct values.
+
+        If the correct substring is an empty string (i.e. the substring
+        should not be present), the invalid substring is placed at
+        the end of the value, otherwise it replaces the correct value.
+
+        Parameters
+        ----------
+        column: TestDataSet
```
Comment (alessiamarcolini) [inline tests/test_dataset.py:690]:
```suggestion
        column : TestDataSet
```

File: tests/test_dataset.py [hunk 15]
```diff
@@ -213,25 +213,43 @@ def __init__(self, sample_size: int = None):
...
-        if column.name in self._name_to_index_map:
-            raise ValueError(
-                "The name and/or the column id is already present"
-                " in the Dataset. If this is expected, set ``overwrite`` to True"
-            )
+
+        Parameters
+        ----------
+        column: Union[TestColumn, _TestColumn]
+            Column that needs to be validated. If ``is_updating`` is True,
```
Comment (alessiamarcolini) [inline tests/test_dataset.py:230]: this is not entirely clear to me

**Review by alessiamarcolini — CHANGES_REQUESTED** [review]
for the first round just a bunch of stylistic issues (that are easier to spot during the first read)

File: tests/test_dataset.py [hunk 16]
```diff
@@ -0,0 +1,741 @@
...
+    """
+    Find the indexes of equally spaced elements in a list following the rule "id=A*x+b"
+
+    In a list with ``list_length`` elements, this function will find the indexes
+    of equally spaced elements. These element indexes will follow the rule
+            "element_id = ``slope`` * id_counter + ``bias``"
+    with "id_counter" in 0 .. ``element_id_count``.
+    The function also takes into account that the list length is finite, and when
+    element_id > ``list_length`` the element_id will be calculated at the beginning
+    of the list, as if the list was a loop. This allows the function to always
```
Comment (alessiamarcolini) [inline tests/test_dataset.py:25]:
```suggestion
    of the list, as if the list was a circular list. This allows the function to always
```

File: tests/test_dataset.py [hunk 17]
```diff
@@ -0,0 +1,741 @@
...
+    return ``element_id_count`` elements.
```
Comment (alessiamarcolini) [inline tests/test_dataset.py:26]: missing Parameters and Return section

File: tests/test_dataset.py [hunk 18]
```diff
@@ -0,0 +1,741 @@
...
+    """
+    element_ids = []
+    for id_counter in range(element_id_count):
+        element_id = slope * id_counter + bias
+        if element_id > list_length:
+            element_ids.append(element_id - list_length)
+        else:
+            element_ids.append(element_id)
```
Comment (alessiamarcolini) [inline tests/test_dataset.py:34]:
```suggestion
        element_ids.append(element_id % list_length)
```

File: tests/test_dataset.py [hunk 19]
```diff
@@ -0,0 +1,741 @@
...
+
+
+@dataclass
+class TestColumn:
+    """
+    DataClass used to initialize columns in TestDataSet class.
+
+    Parameters
+    ----------
+    name: str
```
Comment (alessiamarcolini) [inline tests/test_dataset.py:46]:
```suggestion
    name : str
```

File: tests/test_dataset.py [hunk 20]
```diff
@@ -0,0 +1,741 @@
...
+        Name of the column.
+    original_values: Union[pd.Series, np.ndarray, List, Tuple]
```
Comment (alessiamarcolini) [inline tests/test_dataset.py:48]:
```suggestion
    original_values : Union[pd.Series, np.ndarray, List, Tuple]
```

File: tests/test_dataset.py [hunk 21]
```diff
@@ -0,0 +1,741 @@
...
+        Values of the column.
+    dtype: Union[type, str]
```
Comment (alessiamarcolini) [inline tests/test_dataset.py:50]:
```suggestion
    dtype : Union[type, str]
```

File: tests/test_dataset.py [hunk 22]
```diff
@@ -0,0 +1,741 @@
...
+            ReverseFeatureOperation).
+        value_to_fix: Any
+            Value that is supposed to be found after that the proper correction
+            is applied to ``value_to_fix`` argument.
+        """
+        self._values_to_fix[index] = value_to_fix
+        self._values_after_fix[index] = value_after_fix
+
+    def update_column_values(
+        self, values_to_fix: Iterable[Any], values_after_fix: Iterable[Any]
```
Comment (alessiamarcolini) [inline tests/test_dataset.py:186]: why `Iterable` instead of `Sequence` maybe?

File: tests/test_dataset.py [hunk 23]
```diff
@@ -0,0 +1,741 @@
...
+        values_to_fix: Iterable[Any]
+            Values that are supposed to be found after that the proper correction
+            is applied to ``values_to_fix`` argument.
+        """
+        self._values_to_fix = values_to_fix
+        self._values_after_fix = values_after_fix
+
+
+class TestDataSet:
+    def __init__(self, sample_size: int = None):
```
Comment (alessiamarcolini) [inline tests/test_dataset.py:205]:
```suggestion
    def __init__(self, sample_size: Optional[int] = None):
```

File: tests/test_dataset.py [hunk 24]
```diff
@@ -0,0 +1,741 @@
...
+        # Creating two dicts with the same elements (passed by address), so that
+        # retrieving an element by name or index takes the same time
+        self._columns_by_index = []
+        self._sample_size = sample_size
+        self._name_to_index_map = {}
+
+    @property
+    def sample_size(self) -> Union[int, None]:
```
Comment (alessiamarcolini) [inline tests/test_dataset.py:213]:
```suggestion
    def sample_size(self) -> Optional[int]:
```

File: tests/test_dataset.py [hunk 25]
```diff
@@ -0,0 +1,741 @@
...
+        The returned dataset is modified along with ``dataframe_to_fix`` to keep
+        track of the values inserted by the ReverseFeatureOperation instances
+        applied to columns.
+        Since the process of properly correcting ``dataframe_to_fix`` cannot always
+        be fully reversible, these are the values that the ``dataframe_to_fix`` dataset
+        will contain after that process.
+
+        Returns
+        -------
+        pd.Series
```
Comment (alessiamarcolini) [inline tests/test_dataset.py:339]: not compliant with the return statement

File: tests/test_dataset.py [hunk 26]
```diff
@@ -0,0 +1,741 @@
...
+        # retrieving an element by name or index takes the same time
+        self._columns_by_index = []
+        self._sample_size = sample_size
+        self._name_to_index_map = {}
+
+    @property
+    def sample_size(self) -> Union[int, None]:
+        return self._sample_size
+
+    def _validate_testcolumn(
```
Comment (alessiamarcolini) [inline tests/test_dataset.py:216]: I would change this in a static method

File: tests/test_dataset.py [hunk 27]
```diff
@@ -0,0 +1,741 @@
...
+        value: _TestColumn
+            New _TestColumn instance that replaces the column identified
+            by ``column_name_id``.
+        """
+        self._validate_testcolumn(value, is_updating=True)
+
+        if isinstance(column_name_id, int):
+            self._columns_by_index[column_name_id] = value
+        else:
+            self._columns_by_index[self._name_to_index_map[str(column_name_id)]] = value
```
Comment (alessiamarcolini) [inline tests/test_dataset.py:374]: why converting `column_name_id` to str?

File: tests/test_dataset.py [hunk 28]
```diff
@@ -0,0 +1,741 @@
...
+        elif isinstance(column_name_id, int):
+            return column_name_id < len(self._columns_by_index)
+        else:
+            return False
+
+    def __len__(self):
+        """Return number of columns in this instance"""
+        return len(self._columns_by_index)
+
+    def shape(self) -> Tuple[Union[int, None], int]:
```
Comment (alessiamarcolini) [inline tests/test_dataset.py:396]:
```suggestion
    def shape(self) -> Tuple[Optional[int], int]:
```

File: tests/test_dataset.py [hunk 29]
```diff
@@ -0,0 +1,741 @@
...
+        also keeps track of the related correction (because some Operations like
+        inserting NaNs are not fully reversible). For this reason ``column`` keeps
+        track of the original raw values (simulated by this ReverseFeatureOperation)
+        and of the values that are supposed to be found after the corresponding
+        FeatureOperation is applied.
+        """
+        raise NotImplementedError
+
+
+class Compose(ReverseFeatureOperation):
```
Comment (alessiamarcolini) [inline tests/test_dataset.py:471]: missing docstring

File: tests/test_dataset.py [hunk 30]
```diff
@@ -0,0 +1,741 @@
...
+import numpy as np
+import pandas as pd
+
+INVALID_NAN_BIAS = 0
+INVALID_STRING_BIAS = 1
+INVALID_SUBSTRING_BIAS = 2
+
+
+def _sample_index_in_list(
+    element_id_count: int, slope: int, bias: int, list_length: int
```
Comment (alessiamarcolini) [inline tests/test_dataset.py:14]: to me `element_id_count ` is not an intuitive name

File: tests/test_dataset.py [hunk 31]
```diff
@@ -0,0 +1,741 @@
...
+            nan_sample_ids = _sample_index_in_list(
+                element_id_count=self._error_count,
+                slope=dataset.sample_size // self._error_count,
+                # This BIAS prevents the function to insert invalid strings for the same
+                # samples/ids where it may insert other type of errors
+                bias=column.col_id + INVALID_NAN_BIAS,
+                list_length=dataset.sample_size,
+            )
+
+            # For different sample_id, replacing different elements from the list of
```
Comment (alessiamarcolini) [inline tests/test_dataset.py:533]: this comment is odd, copy-paste leftover

File: tests/test_dataset.py [hunk 32]
```diff
@@ -0,0 +1,741 @@
...
+        error_count: int
+            Number of invalid strings that are inserted in column.
+        replacement_map: Dict
+            Dictionary where the keys are the invalid string that will replace
+            some elements of ``values`` argument. The dictionary values are the
+            values that will replace those invalid strings in the `correct_column`
+            property.
+        """
+        super().__init__(column_names=column_names)
+        self._error_count = error_count
```
Comment (alessiamarcolini) [inline tests/test_dataset.py:575]: why here you don't validate the `replacement_map`?

File: tests/test_dataset.py [hunk 33]
```diff
@@ -0,0 +1,741 @@
...
+        self._columns_by_index = []
+        self._sample_size = sample_size
+        self._name_to_index_map = {}
+
+    @property
+    def sample_size(self) -> Union[int, None]:
+        return self._sample_size
+
+    def _validate_testcolumn(
+        self, column: Union[TestColumn, _TestColumn], is_updating: bool = False
```
Comment (alessiamarcolini) [inline tests/test_dataset.py:217]: you could change `is_updating` with `existing_column`

Comment (lorenz-gorini) [reply tests/test_dataset.py]: I split the two methods since they were different except when checking the sample size

**Comment by alessiamarcolini** [comment]
@lorenz-gorini please be aware that the code is in both files now

**Review by lorenz-gorini — COMMENTED** [review]
> @lorenz-gorini please be aware that the code is in both files now

Solved that. I did not pull right after the commit from github comments and the file rename created two files when pulling

File: tests/datasim.py [hunk 34]
```diff
@@ -0,0 +1,833 @@
...
+                "Dataset. If an update is required, use __setitem__ by indexing"
+            )
+        if len(column) != self.sample_size:
+            raise ValueError(
+                f'`column` argument named "{column.name}" has {len(column)}'
+                f" values, while the other dataset columns have {self.sample_size}"
+                " values."
+            )
+
+    def _validate_testcolumn_to_update(self, column: _TestColumn):
```
Comment (alessiamarcolini) [inline tests/datasim.py:274]: I would have preferred to have one function, now 2/3 of the code inside is duplicated...

File: tests/datasim.py [hunk 35]
```diff
@@ -0,0 +1,833 @@
...
+            self._columns_by_index[self._name_to_index_map[column_name_id]] = value
+
+    def __contains__(self, column_name_id: Union[int, str]) -> bool:
+        """
+        Check if the column identified by ``column_name_id`` is present in TestDataSet
+
+        Parameters
+        ----------
+        column_name_id : Union[int, str]
+            Name or ID of the column whose presence needs to be checked.
```
Comment (alessiamarcolini) [inline tests/datasim.py:456]: missing Returns section

File: tests/test_dataset.py (hunk 26 above)
Comment (alessiamarcolini) [reply tests/test_dataset.py]: alright my bad

**Review by alessiamarcolini — APPROVED** [review]

**Pull request merged**